packages/react/src/components/cookieConsent/modal/Modal.tsx

```diff
@@ -0,0 +1,74 @@
+import React, { useContext, useEffect, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { VisuallyHidden } from '@react-aria/visually-hidden';
+
+import classNames from '../../../utils/classNames';
+import styles from '../CookieConsent.module.scss';
+import { CookieConsentContext, useCookieConsentUiTexts, useFocusShift } from '../CookieConsentContext';
+import { Content } from '../content/Content';
+import { useEscKey } from '../useEscKey';
+
+export function Modal(): React.ReactElement | null {
+  const cookieConsentContext = useContext(CookieConsentContext);
```

Comment:
This might cause lot of unnecessary re-renders of the whole modal component. If you check any checkbox in the cookie modal, then the whole modal will be re-rendered again when I tested. This will cause lot of unnecessary JS to be run. For example, the ReactDOM.createPortal will be run again, and all the initial JS that are inside this component will be run again.

Reply:
React Context is commonly used in React Apps instead of prop drilling. With context the data is in one place data and also actions that change data, like button clicks etc. Without the context, passing translations, consents and other data would require Redux, MobX or passing down objects from parent to grand-grand-grand-grand-children.

Because of context, the root component is re-rendered every time user does something. But that happens quite rarely - maybe once in 5 seconds for a fast user. Usually user just closes the modal with a single button click. Some optimization/memoizing can be applied. Mainly for the buttons, because they never change. 

One checkbox click can affect multiple components. Optimizing changes that have multiple side-effects may not be worth the effort when the component re-renders so rarely.

Re-rendering does not always update the DOM, so running functional components is not slow.

I have no experience with the Portal, but recreating that on each context change seems excessive. I would remove the portal usage to another component. if possible.

Reply:
Good points! 

> One checkbox click can affect multiple components. Optimizing changes that have multiple side-effects may not be worth the effort when the component re-renders so rarely.

Do you mean components that are outside of the modal? They could use the cookie consent context, and yeah that way clicking a checkbox could affect multiple components outside the modal. But inside the modal not that many components change after clicking the checkbox, or have I understood it wrong? But I think you are right, if it's hard to optimize it might not be worth it 🤔 

> Re-rendering does not always update the DOM, so running functional components is not slow.

True, but it still will run unnecessary javascript.

Reply:
I mean clicking a group's checkbox will affect checkboxes in its child consents and their checkboxes. So one click must update state of the child components too. Splitting context will solve that, but still the state of all checkboxes must be known to the controller which writes the cookies. Optimizing state is very hard in this kind of component where all children affect the "root state"

Adding React.Memo to optional consents and its every consent group can prevent some re-rendering.  Required consents cannot be changed, so React.Memo can be used there too.

Reply:
Tried a bit of memoizing on some of the components. But... User actions update the context. Context updates causes components to re-render. So, memoization is useless unless context is removed from the render process.

What could be done:
- split texts in to their own context which will change only when language is changed. This won't actually decrease re-renderings, because about all components which use texts also use other data from the context. But anyway some optimisation is gained.
- memoize all functions which use consentController. ConsentController itself is memoized, so functions calling it can also be memoized. This way some components can be memoized. For example buttons.
- split components into two: a part which uses context and a part which only creates elements . Latter part might be memoizable.